compliance_checker/cf/cf_18.py

```diff
@@ -132,7 +140,189 @@ def check_taxa(self, ds: MemoizedDataset):
+                                                          standard_name="biological_taxon_identifier")
+
+        def match_taxa_standard_names(standard_name_string):
+            """
+            Match variables which are standard_names related to taxa, but
+            are not the taxon identifiers or LSIDs themselves.
+            """
+            return ("taxon" in standard_name_string and
+                    # exclude the identifiers we just looked at
+                    standard_name_string not in {"biological_taxon_identifier",
+                                                 "biological_taxon_name"} and
+                    standard_name_string in self._std_names)
+
+
+        taxa_quantifier_variables = ds.get_variables_by_attributes(
+                                                standard_name=match_taxa_standard_names)
+        # If there are no matches, there either are no taxa variables
+        # or the standard names are not appropriate, which will be picked up
+        # by the standard_name check
+        if not taxa_quantifier_variables:
+            return
+
+
+        for taxon_quantifier_variable in taxa_quantifier_variables:
+            valid_taxa = TestCtx(BaseCheck.HIGH, self.section_titles["6.1.2"])
+            if not isinstance(getattr(taxon_quantifier_variable,
+                                      "coordinates", None), str):
+                valid_taxa.add_failure(f"{taxon_quantifier_variable.name} must have a string valued \"coordinates\" attribute")
+                continue
+
+            #coord_var_names = taxon_quantifier_variable.coordinates.split(" ")
+            coordinate_var_names = (taxon_quantifier_variable.coordinates
+                                    .split(" "))
+            invalid_coord_vars = set(coordinate_var_names) - ds.variables.keys()
+            if invalid_coord_vars:
+                valid_taxa.add_failure("The following values for \"coordinates\" attributes were not found in the dataset's variables "
+                                       f"{invalid_coord_vars}")
+
+            if len(coordinate_var_names) > 2:
+                valid_taxa.add_failure("coordinates attribute for taxon data must either reference one or two varaible name")
+                continue
+
+            coordinate_vars = [ds.variables[var_name] for var_name in
+                                coordinate_var_names]
+
+
+            coord_var_standard_names = {var:
+                                        getattr(var, 'standard_name', None)
+                                        for var in coordinate_vars}
+            #coord_vars = reference_attr_variables(ds, "coordinates", " ")
+            # if we have no authority, we can't check validity of the name -- assume it's OK
+            standard_name_set = set(coord_var_standard_names.values())
+            if set(coord_var_standard_names.keys()) == {"biological_taxon_name"}:
+                # TODO: Check for at least binomial nomenclature?
```

Comment:
Should WoRMS/ITIS still be checked if a variable with "biological_taxon_lsid" is not specified?   The CF Standard Name Table entry mentions in the entry for `biological_taxon_name`:

> "Biological taxon" is a name or other label identifying an organism or a group of organisms as belonging to a unit of classification in a hierarchical taxonomy. The quantity with standard name biological_taxon_name is the human-readable label for the taxon such as Calanus finmarchicus. The label should be registered in either WoRMS (http://www.marinespecies.org) or ITIS (https://www.itis.gov/) and spelled exactly as registered. See Section 6.1.2 of the CF convention (version 1.8 or later) for information about biological taxon auxiliary coordinate variables.

However, the CF interpretation at https://cfconventions.org/Data/cf-conventions/cf-conventions-1.8/cf-conventions.html#taxon-names-and-identifiers seems a little less strict:

> It is strongly recommended in CF that the authority chosen is World Register of Marine Species (WoRMS) for oceanographic data and Integrated Taxonomic Information System (ITIS) for freshwater and terrestrial data. WoRMS LSIDs are built from the WoRMS AphiaID taxon identifier such as "urn:lsid:marinespecies.org:taxname:104464" for AphiaID 104464. This may be converted to a URL by adding prefixes such as ​http://www.lsid.info/. ITIS LSIDs are built from the ITIS Taxonomic Serial Number (TSN), such as "urn:lsid:itis.gov:itis_tsn:180543".

> The biological_taxon_name auxiliary coordinate variable included for human readability is mandatory. The biological_taxon_lsid auxliary coordinate variable included for software agent readability is optional, but strongly recommended. If both are present then each biological_taxon_name coordinate must exactly match the name resolved from the biological_taxon_lsid coordinate. If LSIDs are available for some taxa in a dataset then the biological_taxon_lsid auxiliary coordinate variable should be included and missing data given for those taxa that do not have an identifier.

Reply:
Should we also search for the value(s) of `biological_taxon_name` in WoRMS and ITIS if `biological_taxon_lsid` is not present?

Reply:
I think so... But, keep in mind it has to be spelled exactly as registered in ITIS or WoRMS. 

Also wondering how computationally expensive this check could be. Thinking of the case where there's hundreds of species to look up in both services. 

Maybe this [global names resolver](http://resolver.globalnames.org/api) example can help? https://gist.github.com/adyork/5983a1ed9763ffb37fb7cd15df24e895#file-match_names_gnrd-ipynb

Reply:
The API looks interesting, especially the fuzzy matching.  I mentioned this as a possible desirable feature for CF standard names in #806.  The API looks to be alpha based upon the link, but I'll keep an eye on it.  In general, caching might be possible for web requests.  On one of these I also saw that there was a database dump, but it was about 173 MiB compressed, so that's rather a rather large file, given that many users aren't even checking bio data.